File: src/Heaps.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "Heaps.h"
#include "ladder.h"
#include "Zaks.h"
#include "utilities.h"
#include "LinkedListAPI.h"

void heaps(List *perms, int *perm, int n, int idx)
{
    if (idx == 0)
    {
        //appendPerm(perms, perm, n);
	printPerm(perm, n);
        if(perms->length == 0)
            insertBack(perms, copyIntArr(perm, n));
        return;
    }
    else
    {

        forall(idx)
        {
            heaps(perms, perm, n, idx - 1);
            if (idx % 2 == 0)
            {
                int temp = perm[idx];
                perm[idx] = perm[0];
                perm[0] = temp;
            }
            else
            {
                int temp = perm[idx];
                perm[idx] = perm[x];
                perm[x] = temp;
            }
            insertBack(perms, copyIntArr(perm, n));
            //printPerm(perm, n);
            
        }
        if (idx > 1)
        {
            heaps(perms, perm, n, idx - 1);

        }
    }
}

/**int main()
{
	int* perms = calloc(4, sizeof(int));
	List* permList = initializeList(dummy_print, dummy_delete, dummy_compare);
	int n   = 4;
	int size = n-1; 
	forall(n)
	{
	  perms[x] = x+1;
	}
	heaps(permList, perms, n, size);
}**/

File: test/Heaps.py



def Heaps(k, perm):
    if k == 1:
        print(perm)
        return

    else:
        for i in range(k):
            Heaps(k-1, perm)
            if k %2 == 0:
                perm[i], perm[k-1] = perm[k-1], perm[i]
            else:
                perm[0], perm[k-1] = perm[k-1], perm[0]



if __name__ == "__main__":
    
    l = []
    for i in range(3):
        l.append(i+1)

    Heaps(len(l), l)


